APCAppCore/APCAppCore/UI/Model/APCOnboardingManager.m

```diff
@@ -0,0 +1,238 @@
+
+@end
+
+
+@implementation APCOnboardingManager
+
+
+- (instancetype)initWithProvider:(id<APCOnboardingManagerProvider>)provider user:(APCUser * __nonnull)user {
+    if ((self = [super init])) {
+        self.provider = provider;
+        self.user = user;
+        _signInSupported = YES;
+    }
+    return self;
+}
+
++ (instancetype)managerWithProvider:(id<APCOnboardingManagerProvider>)provider user:(APCUser * __nonnull)user {
+    return [[self alloc] initWithProvider:provider user:user];
+}
+
+
+#pragma mark - Onboarding
+
+- (void)instantiateOnboardingForType:(APCOnboardingTaskType)type {
+    if (self.onboarding) {
+        self.onboarding.delegate = nil;
+        self.onboarding = nil;
+    }
+    self.onboarding = [[APCOnboarding alloc] initWithDelegate:self taskType:type];
+}
+
+- (void)userDidConsentWithResult:(ORKConsentSignatureResult *)consentResult sharingScope:(APCUserConsentSharingScope)sharingScope {
+    NSString *signatureName = [consentResult.signature.givenName stringByAppendingFormat:@" %@",consentResult.signature.familyName];
+    NSData *signatureImage = UIImagePNGRepresentation(consentResult.signature.signatureImage);
+    
+    _user.consentSignatureName = signatureName;
+    _user.consentSignatureImage = signatureImage;
+    _user.consentSignatureDate = consentResult.startDate;
+    _user.sharingScope = sharingScope;
+    _user.userConsented = YES;
+}
+
+- (void)userDeclinedConsent {
+    [[NSNotificationCenter defaultCenter] postNotificationName:APCUserDidDeclineConsentNotification object:self];
+}
+
+- (void)onboardingDidFinish {
+    [self completeOnboardingAsSignIn:(self.onboarding.taskType == kAPCOnboardingTaskTypeSignIn)];
+}
+
+- (void)onboardingDidFinishAsSignIn {
+    [self completeOnboardingAsSignIn:YES];
+}
+
+- (void)completeOnboardingAsSignIn:(BOOL)wasSignIn {
+    if (wasSignIn) {
+        self.user.signedIn = YES;
+        //[(APCAppDelegate *)[UIApplication sharedApplication].delegate afterOnBoardProcessIsFinished];     // TODO: Untwine this spaghetti
+        [[NSNotificationCenter defaultCenter] postNotificationName:APCUserSignedInNotification object:self];
+    } else {
+        self.user.signedUp = YES;
+        [[NSNotificationCenter defaultCenter] postNotificationName:APCUserSignedUpNotification object:self];
+    }
+    
+    [[NSNotificationCenter defaultCenter] postNotificationName:APCUserDidConsentNotification object:self];
+}
+
+
+#pragma mark - Permissions
+
+- (APCPermissionsManager *)permissionsManager {
+    if (!_permissionsManager) {
+        _permissionsManager = [self createPermissionsManager];
+    }
+    return _permissionsManager;
+}
+
+- (APCPermissionsManager *)createPermissionsManager {
+    APCPermissionsManager *manager = [APCPermissionsManager new];
+    
+    // set HealthKit characteristic types to read
+    NSMutableArray *types = [NSMutableArray arrayWithCapacity:3];
+    for (NSNumber *type in self.userProfileElements) {
```

Comment:
The intention with the manager is that you create a subclass, configure everything in the subclass and let your app delegate (which doesn't need to be an `APCAppDelegate` subclass) instantiate it.

(The array limit 3 is just to optimize mutable array initialization, it can still grow larger than 3)